File: learn_bigdata_flink/src/main/java/com/reiser/xeye/adclick/AdClickCount.java
package com.reiser.xeye.adclick;

import org.apache.flink.api.common.functions.AggregateFunction;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.common.state.ValueState;
import org.apache.flink.api.common.state.ValueStateDescriptor;
import org.apache.flink.api.java.functions.KeySelector;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.AssignerWithPeriodicWatermarks;
import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
import org.apache.flink.streaming.api.functions.windowing.WindowFunction;
import org.apache.flink.streaming.api.watermark.Watermark;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
import org.apache.flink.util.Collector;
import org.apache.flink.util.OutputTag;

import javax.annotation.Nullable;

/**
 * @author: reiserx
 * Date:2020/11/8
 * Des:广告点击统计
 * 1.一天之内点击超过100次的 进入黑名单
 * 2.每隔5秒统计最近1小时的各省份的广告点击
 */
public class AdClickCount {

    static OutputTag<BlackListWarning> mBlackListWarningOutputTag = new OutputTag<BlackListWarning>("blackwaring") {
    };


    public static void main(String[] args) throws Exception {

        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);
        //注册时间类型为事件时间
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);


        //1.获取数据
        SingleOutputStreamOperator<AdClickEvent> sourceStream = env.readTextFile("/Users/reiserx/code/bigdata/learn_bigdata/learn_bigdata_flink/src/main/resources/data3.csv").map(new MapFunction<String, AdClickEvent>() {
            @Override
            public AdClickEvent map(String line) throws Exception {
                String[] fields = line.split(",");
                return new AdClickEvent(Long.parseLong(fields[0]), Long.parseLong(fields[1]), fields[2], fields[3], Long.parseLong(fields[4]));
            }
        }).assignTimestampsAndWatermarks(new AdClickEventTimeExtractor());
        //计算黑名单
        SingleOutputStreamOperator<AdClickEvent> filterBlackStream = sourceStream.keyBy(new KeySelector<AdClickEvent, Tuple2<Long, Long>>() {
            @Override
            public Tuple2<Long, Long> getKey(AdClickEvent value) throws Exception {
                return Tuple2.of(value.getUserId(), value.getAdId());
            }
        }).process(new FilterBlackListUser(100));

        filterBlackStream.getSideOutput(mBlackListWarningOutputTag).print();

        // 说去各省的广告点击
        filterBlackStream
                .keyBy(AdClickEvent::getProvince)
                .timeWindow(Time.minutes(10), Time.seconds(5))
                .aggregate(new AdClickEventCount(), new AdClickWindow())
                .print();

        env.execute(AdClickCount.class.getSimpleName());

    }

    static class AdClickWindow implements WindowFunction<Long, CountByProvince, String, TimeWindow> {

        @Override
        public void apply(String key, TimeWindow window, Iterable<Long> input, Collector<CountByProvince> out) throws Exception {
            out.collect(new CountByProvince(String.valueOf(window.getEnd()), key, input.iterator().next()));


        }
    }

    static class AdClickEventCount implements AggregateFunction<AdClickEvent, Long, Long> {
        @Override
        public Long createAccumulator() {
            return 0L;
        }

        @Override
        public Long add(AdClickEvent value, Long accumulator) {
            return accumulator + 1;
        }

        @Override
        public Long getResult(Long accumulator) {
            return accumulator;
        }

        @Override
        public Long merge(Long a, Long b) {
            return a + b;
        }
    }

    static class AdClickEventTimeExtractor implements AssignerWithPeriodicWatermarks<AdClickEvent> {
        long maxEventTimestamps = 0;
        long maxOutputOrderness = 10000L;

        @Nullable
        @Override
        public Watermark getCurrentWatermark() {
            return new Watermark(maxEventTimestamps - maxOutputOrderness);
        }

        @Override
        public long extractTimestamp(AdClickEvent element, long previousElementTimestamp) {
            long currentEventTimestamps = element.getTimestamp() * 1000;
            maxEventTimestamps = Math.max(currentEventTimestamps, maxEventTimestamps);
            return currentEventTimestamps;
        }
    }


    private static class FilterBlackListUser extends KeyedProcessFunction<Tuple2<Long, Long>, AdClickEvent, AdClickEvent> {
        private ValueState<Long> mCountState;
        private ValueState<Long> mTimeState;
        private ValueState<Boolean> mInBlackListState;
        private int mMaxCount;

        FilterBlackListUser(int maxCount) {
            this.mMaxCount = maxCount;
        }


        @Override
        public void open(Configuration parameters) throws Exception {
            super.open(parameters);

            mCountState = getRuntimeContext().getState(new ValueStateDescriptor<>("count-state", Long.class));
            mTimeState = getRuntimeContext().getState(new ValueStateDescriptor<>("time-state", Long.class));
            mInBlackListState = getRuntimeContext().getState(new ValueStateDescriptor<>("black-state", Boolean.class));
        }

        @Override
        public void processElement(AdClickEvent value, Context ctx, Collector<AdClickEvent> out) throws Exception {


            Long count = mCountState.value();
            if (count == null) {
                long time = (ctx.timerService().currentProcessingTime() / (1000 * 60 * 60 * 24) + 1) * (1000 * 60 * 60 * 24);
                mTimeState.update(time);
                ctx.timerService().registerEventTimeTimer(time);
                count = 0L;
            }

            count++;
            if (count >= mMaxCount) {
                if(mInBlackListState.value() ==null) {
                    mInBlackListState.update(true);
                    ctx.output(mBlackListWarningOutputTag, new BlackListWarning(value.getUserId(), value.getAdId(), "Click over" + mMaxCount + " times"));
                }
                return;
            }
            mCountState.update(count);
            out.collect(value);
        }

        @Override
        public void onTimer(long timestamp, OnTimerContext ctx, Collector<AdClickEvent> out) throws Exception {
            super.onTimer(timestamp, ctx, out);
            if (timestamp == mTimeState.value()) {
                mTimeState.clear();
                mCountState.clear();
            }
        }
    }


    static class AdClickEvent {
        private Long userId;
        private Long adId;
        private String province;
        private String city;
        private Long timestamp;

        public AdClickEvent(Long userId, Long adId, String province, String city, Long timestamp) {
            this.userId = userId;
            this.adId = adId;
            this.province = province;
            this.city = city;
            this.timestamp = timestamp;
        }

        public Long getUserId() {
            return userId;
        }

        public void setUserId(Long userId) {
            this.userId = userId;
        }

        public Long getAdId() {
            return adId;
        }

        public void setAdId(Long adId) {
            this.adId = adId;
        }

        public String getProvince() {
            return province;
        }

        public void setProvince(String province) {
            this.province = province;
        }

        public String getCity() {
            return city;
        }

        public void setCity(String city) {
            this.city = city;
        }

        public Long getTimestamp() {
            return timestamp;
        }

        public void setTimestamp(Long timestamp) {
            this.timestamp = timestamp;
        }
    }

    static class CountByProvince {
        private String windowEnd;
        private String province;
        private Long count;

        public CountByProvince(String windowEnd, String province, Long count) {
            this.windowEnd = windowEnd;
            this.province = province;
            this.count = count;
        }

        public String getWindowEnd() {
            return windowEnd;
        }

        public void setWindowEnd(String windowEnd) {
            this.windowEnd = windowEnd;
        }

        public String getProvince() {
            return province;
        }

        public void setProvince(String province) {
            this.province = province;
        }

        public Long getCount() {
            return count;
        }

        public void setCount(Long count) {
            this.count = count;
        }

        @Override
        public String toString() {
            return "CountByProvince{" +
                    "windowEnd='" + windowEnd + '\'' +
                    ", province='" + province + '\'' +
                    ", count=" + count +
                    '}';
        }
    }

    public static class BlackListWarning {
        private Long userId;
        private Long adId;
        private String msg;

        public BlackListWarning(Long userId, Long adId, String msg) {
            this.userId = userId;
            this.adId = adId;
            this.msg = msg;
        }

        public Long getUserId() {
            return userId;
        }

        public void setUserId(Long userId) {
            this.userId = userId;
        }

        public Long getAdId() {
            return adId;
        }

        public void setAdId(Long adId) {
            this.adId = adId;
        }

        public String getMsg() {
            return msg;
        }

        public void setMsg(String msg) {
            this.msg = msg;
        }

        @Override
        public String toString() {
            return "BlackListWarning{" +
                    "userId=" + userId +
                    ", adId=" + adId +
                    ", msg='" + msg + '\'' +
                    '}';
        }
    }
}

File: learn_bigdata_flink/src/main/scala/com/reiser/yeye/adclick/AdClickCount.scala
package com.reiser.yeye.adclick

import java.sql.Timestamp

import org.apache.flink.api.common.functions.AggregateFunction
import org.apache.flink.api.common.state.{ValueState, ValueStateDescriptor}
import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.functions.{AssignerWithPeriodicWatermarks, KeyedProcessFunction}
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.scala.function.WindowFunction
import org.apache.flink.streaming.api.watermark.Watermark
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.streaming.api.windowing.windows.TimeWindow
import org.apache.flink.util.Collector


// 输入的广告点击事件样例类
case class AdClickEvent( userId: Long, adId: Long, province: String, city: String, timestamp: Long )
// 按照省份统计的输出结果样例类
case class CountByProvince( windowEnd: String, province: String, count: Long )
// 输出的黑名单报警信息
case class BlackListWarning( userId: Long, adId: Long, msg: String )


/**
 * 广告点击统计
 */
object AdClickCount {

  private val outputBlackList = new OutputTag[BlackListWarning]("blackwaring")

  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(4)
    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)

    //1. 获取数据
    val adEventStream = env.readTextFile("/Users/reiserx/code/bigdata/learn_bigdata/learn_bigdata_flink/src/main/resources/data3.csv").map(data => {
      val dataArray = data.split(",")
      AdClickEvent(dataArray(0).trim.toLong, dataArray(1).trim.toLong, dataArray(2).trim, dataArray(3).trim, dataArray(4).trim.toLong)
    })
      .assignTimestampsAndWatermarks(new AdClickEventTimeExtractor())

    //2. 计算黑名单
    val filterBlackListStream = adEventStream.keyBy(data => (data.userId, data.adId))
      .process(new FilterBlackListUser(100))

    //3. 打印黑名单
    filterBlackListStream.getSideOutput(outputBlackList)
      .print()

    //4.计算各省份的广告点击量

    val adCountStream = filterBlackListStream.keyBy(_.province)
      .timeWindow(Time.hours(1), Time.seconds(5))
      .aggregate(new AdClickCount, new AdClickWindow)

    adCountStream.print()


    env.execute("AdClickCount")

  }

  /**
   * 自定义窗口处理
   */
  class AdClickWindow extends WindowFunction[Long,CountByProvince,String,TimeWindow]{
    override def apply(key: String, window: TimeWindow,
                       input: Iterable[Long],
                       out: Collector[CountByProvince]): Unit = {
      out.collect(CountByProvince(new Timestamp(window.getEnd).toString,
        key,input.iterator.next()))
    }
  }

  class AdClickCount extends AggregateFunction[AdClickEvent,Long,Long]{
    override def createAccumulator(): Long = 0L

    override def add(in: AdClickEvent, acc: Long): Long = acc + 1

    override def getResult(acc: Long): Long = acc

    override def merge(acc: Long, acc1: Long): Long = acc + acc1
  }

  /**
   * 过滤黑名单数据
   * @param maxCount
   */
  class FilterBlackListUser(maxCount:Int)
    extends KeyedProcessFunction[(Long,Long),AdClickEvent,AdClickEvent]{
    //保存当前用户对当前广告的点击量
    lazy val countState:ValueState[Long] = getRuntimeContext.getState(
      new ValueStateDescriptor[Long]("count-state",classOf[Long]))


    //保存是否发送黑名单的状态
    lazy val isSetBlackList:ValueState[Boolean] = getRuntimeContext.getState(
      new ValueStateDescriptor[Boolean]("issent-state",classOf[Boolean]))

    //保存定时器触发的时间戳
    lazy val resetTimeer:ValueState[Long] = getRuntimeContext.getState(
      new ValueStateDescriptor[Long]("resettime-state",classOf[Long])
    )

    override def processElement(value: AdClickEvent,
                                ctx: KeyedProcessFunction[(Long, Long),
                                  AdClickEvent, AdClickEvent]#Context,
                                out: Collector[AdClickEvent]): Unit = {
      val currentCount = countState.value()
      //如果是第一次处理，注册定时器，每天00:00触发
      if(currentCount == 0){
        val ts = (ctx.timerService().currentProcessingTime()/(1000*60*60*24) +1) * (1000 * 60 * 60 * 24)
        resetTimeer.update(ts)
        ctx.timerService().registerProcessingTimeTimer(ts)
      }
      //判断计数是否达到上线，如果达到加入黑名单
      if(currentCount >= maxCount){
        if(!isSetBlackList.value()){
          isSetBlackList.update(true)
          //输入到侧输出流
          ctx.output(outputBlackList,
            BlackListWarning(value.userId,value.adId,"Click over" + maxCount +" times"))
        }
        return
      }

      countState.update(currentCount + 1)
      out.collect(value)
    }

    override def onTimer(timestamp: Long,
                         ctx: KeyedProcessFunction[(Long, Long),
                           AdClickEvent, AdClickEvent]#OnTimerContext,
                         out: Collector[AdClickEvent]): Unit = {
      if(timestamp == resetTimeer.value()){
        isSetBlackList.clear()
        countState.clear()
        resetTimeer.clear()
      }

    }

  }

}

class AdClickEventTimeExtractor extends AssignerWithPeriodicWatermarks[AdClickEvent]{

  var currentMaxEventTime = 0L
  val maxOufOfOrderness = 10000 //最大乱序时间

  override def getCurrentWatermark: Watermark = {
    new Watermark(currentMaxEventTime - maxOufOfOrderness)
  }

  override def extractTimestamp(element: AdClickEvent, previousElementTimestamp: Long): Long = {
    //时间字段
    val timestamp = element.timestamp * 1000

    currentMaxEventTime = Math.max(element.timestamp, currentMaxEventTime)
    timestamp;
  }
}
